synapse/handlers/sso.py

```diff
@@ -179,9 +190,19 @@ async def get_mxid_from_sso(
         for i in range(self._MAP_USERNAME_RETRIES):
             try:
                 attributes = await sso_to_matrix_id_mapper(i)
+            except (RedirectException, MappingException):
+                # Mapping providers are allowed to issue a redirect (e.g. to ask
+                # the user for more information) and can issue a mapping exception
+                # if a name cannot be generated.
+                raise
             except Exception as e:
+                # Any other exception is unexpected.
+                logger.error(
+                    "Unexpected error when extracting user attributes from SSO response: %s"
+                    % (e,)
+                )
                 raise MappingException(
-                    "Could not extract user attributes from SSO response: " + str(e)
+                    "Could not extract user attributes from SSO response."
                 )
```

Comment:
given that there will be another ERROR logged later when the Mapping exception is caught, I don't think this will be clear. I don't think we should log twice at ERROR for the same problem.

Instead I would:


```python
raise MappingException(
    "Unexpected error when extracting user attributes from SSO response."
) from e
```

That would get logged as something like:

```
2020-12-04 12:47:43,699 - synapse.handlers.saml_handler - 205 - ERROR - POST-4 - Could not map user
Traceback (most recent call last):
  File "/home/rav/work/synapse/synapse/handlers/sso.py", line 181, in get_mxid_from_sso
    attributes = await sso_to_matrix_id_mapper(i)
  File "/home/rav/work/synapse/synapse/handlers/saml_handler.py", line 262, in saml_response_to_remapped_user_attributes
    saml2_auth, failures, client_redirect_url
  File "/home/rav/work/matrix-synapse-saml-mozilla/matrix_synapse_saml_mozilla/mapping_provider.py", line 162, in saml_response_to_user_attributes
    raise e
synapse.api.errors.RedirectException: 302: Redirect to /_matrix/saml2/pick_username/

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/rav/work/synapse/synapse/handlers/saml_handler.py", line 202, in handle_saml_response
    saml2_auth, relay_state, user_agent, ip_address
  File "/home/rav/work/synapse/synapse/handlers/saml_handler.py", line 304, in _map_saml_response_to_user
    grandfather_existing_users,
  File "/home/rav/work/synapse/synapse/handlers/sso.py", line 185, in get_mxid_from_sso
    ) from e
synapse.handlers.sso.MappingException: Unexpected error when extracting user attributes from SSO response.
```

(modulo the fact we now propagate RedirectExcepitons) ... which I think is adequate?

Reply:
Oh, that's a good idea! I always forget about raising from other exceptions! 👍